source/extensions/filters/http/grpc_web/grpc_web_filter.cc

```diff
@@ -54,6 +74,82 @@ bool GrpcWebFilter::isGrpcWebRequest(const Http::RequestHeaderMap& headers) {
   return false;
 }
 
+bool GrpcWebFilter::isProtoEncodedGrpcWebResponseHeaders(
+    const Http::ResponseHeaderMap& headers) const {
+  // We expect the response headers to have 200 OK status (a valid gRPC, also gRPC-Web, response
+  // needs to have 200 OK status
+  // https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#responses) and contain
+  // proto-encoded gRPC-Web content-type.
+  return Http::Utility::getResponseStatus(headers) == enumToInt(Http::Code::OK) &&
+         hasProtoEncodedGrpcWebContentType(headers);
+}
+
+bool GrpcWebFilter::hasProtoEncodedGrpcWebContentType(
+    const Http::RequestOrResponseHeaderMap& headers) const {
+  const Http::HeaderEntry* content_type = headers.ContentType();
+  if (content_type != nullptr) {
+    absl::string_view content_type_value = content_type->value().getStringView();
+    // We ignore "parameter" value. Note that "*( ";" parameter )" indicates that there can be
+    // multiple parameters.
+    absl::string_view current_content_type =
+        StringUtil::rtrim(content_type_value.substr(0, content_type_value.find_first_of(';')));
+    // We expect only proto encoding response
+    // https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md. And the value of media-type is
+    // case-sensitive https://tools.ietf.org/html/rfc2616#section-3.7.
+    return StringUtil::CaseInsensitiveCompare()(
+               current_content_type, Http::Headers::get().ContentTypeValues.GrpcWebProto) ||
+           StringUtil::CaseInsensitiveCompare()(current_content_type,
+                                                Http::Headers::get().ContentTypeValues.GrpcWeb);
+  }
+  return false;
+}
+
+// If response headers do not contain valid response headers, it needs transformation.
+bool GrpcWebFilter::needsTransformationForNonProtoEncodedResponse(Http::ResponseHeaderMap& headers,
+                                                                  bool end_stream) const {
+  return Runtime::runtimeFeatureEnabled(
+             "envoy.reloadable_features.grpc_web_fix_non_proto_encoded_response_handling") &&
+         // We do transformation when the response headers is not gRPC response headers and it is
+         // not proto-encoded gRPC-Web response headers.
+         !Grpc::Common::isGrpcResponseHeaders(headers, end_stream) &&
+         !isProtoEncodedGrpcWebResponseHeaders(headers);
+}
+
+void GrpcWebFilter::mergeAndLimitNonProtoEncodedResponseData(Buffer::OwnedImpl& output,
+                                                             Buffer::Instance* last_data) {
+  const auto* encoding_buffer = encoder_callbacks_->encodingBuffer();
+  if (encoding_buffer != nullptr) {
+    encoder_callbacks_->modifyEncodingBuffer([&output](Buffer::Instance& buffered) {
+      output.move(buffered, MAX_BUFFERED_PLAINTEXT_LENGTH);
+      buffered.drain(buffered.length());
+    });
+  }
+
+  // In the case of local reply, "encoding_buffer" is nullptr and we only have filled "last_data".
```

Comment:
Wouldn't this also be when the response only contains a single DATA chunk?

Reply:
Added more comments to it.